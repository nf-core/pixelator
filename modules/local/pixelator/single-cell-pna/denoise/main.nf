process PIXELATOR_PNA_DENOISE {
    tag "$meta.id"
    label 'process_high'

    // TODO: Add conda
    // conda "bioconda::pixelator=0.18.2"

    container "${workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container
        ? 'quay.io/pixelgen-technologies/pixelator:0.21.2'
        : 'quay.io/pixelgen-technologies/pixelator:0.21.2'}"

    input:
    tuple val(meta), path(data)

    output:
    tuple val(meta), path("denoise/*.pxl")             , emit: pixelfile
    tuple val(meta), path("denoise/*.report.json")     , emit: report_json
    tuple val(meta), path("denoise/*.meta.json")       , emit: metadata_json
    tuple val(meta), path("denoise/*")                 , emit: all_results
    tuple val(meta), path("*pixelator-denoise.log")    , emit: log

    path "versions.yml"                                 , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:

    prefix = task.ext.prefix ?: "${meta.id}"
    def args = task.ext.args ?: ''

    """
    pixelator \\
        --cores $task.cpus \\
        --log-file ${prefix}.pixelator-denoise.log \\
        --verbose \\
        single-cell-pna \\
        denoise \\
        --output . \\
        $args \\
        $data

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        pixelator: \$(echo \$(pixelator --version 2>/dev/null) | sed 's/pixelator, version //g' )
    END_VERSIONS
    """

    stub:
    prefix = task.ext.prefix ?: "${meta.id}"

    """
    mkdir denoise
    touch denoise/${prefix}.report.json
    touch denoise/${prefix}.meta.json
    touch denoise/${prefix}.pxl
    touch ${prefix}.pixelator-denoise.log

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        pixelator: \$(echo \$(pixelator --version 2>/dev/null) | sed 's/pixelator, version //g' )
    END_VERSIONS
    """
}
