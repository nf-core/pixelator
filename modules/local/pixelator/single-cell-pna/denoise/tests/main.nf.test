nextflow_process {

    name "Test Process PIXELATOR_PNA_DENOISE"
    script "../main.nf"
    process "PIXELATOR_PNA_DENOISE"
    tag "modules"
    tag "pixelator"
    tag "pixelator/pna"
    tag "pixelator/single_cell_pna_denoise"

    test("PNA denoise - small test") {
        when {
            process {
                """
                input[0] = [
                    [ id:'PNA055_Sample07_filtered_S7', design:'pna-2', panel:'proxiome-immuno-155', technology:'pna' ],
                    file(params.pipelines_testdata_base_path + 'testdata/pna/modules/graph/PNA055_Sample07_filtered_S7.graph.pxl', checkIfExists: true),
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out.metadata_json).match("metadata_json") },
                {
                    with (process.out.report_json) {
                        assert snapshot(path(get(0).get(1)).readLines()[0..5]).match("report_json") }
                },
                { assert snapshot(process.out.versions).match("versions") },
                {
                    with (process.out.log) {
                        assert path(get(0).get(1)).readLines().last().contains("Finished pixelator denoise")
                    }
                },
            )
        }
    }
}

