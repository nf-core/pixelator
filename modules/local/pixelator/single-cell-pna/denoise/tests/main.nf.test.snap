{
    "versions": {
        "content": [
            [
                "versions.yml:md5,75e9df3e1d5df7d1b9b4c6d91dcf24eb"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-06-27T14:53:31.633773358"
    },
    "report_json": {
        "content": [
            [
                "{",
                "    \"sample_id\": \"PNA055_Sample07_filtered_S7\",",
                "    \"product_id\": \"single-cell-pna\",",
                "    \"report_type\": \"denoise\",",
                "    \"number_of_umis_removed\": 599,",
                "    \"ratio_of_umis_removed\": 0.019271604143877484,"
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-06-27T14:53:31.61894306"
    },
    "metadata_json": {
        "content": [
            [
                [
                    {
                        "id": "PNA055_Sample07_filtered_S7",
                        "design": "pna-2",
                        "panel": "proxiome-immuno-155",
                        "technology": "pna"
                    },
                    "PNA055_Sample07_filtered_S7.meta.json:md5,d0728a1efc25964039497c7bc22cdc35"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.2"
        },
        "timestamp": "2025-06-27T14:53:31.58914569"
    }
}